llvm/include/llvm/CodeGen/SelectionDAGNodes.h

```diff
@@ -837,29 +835,72 @@ END_TWO_BYTE_PACK()
     }
   };
 
+  class user_iterator {
+    friend class SDNode;
+    use_iterator UI;
+
+    explicit user_iterator(SDUse *op) : UI(op) {};
+
+  public:
+    using iterator_category = std::forward_iterator_tag;
+    using value_type = SDNode *;
+    using difference_type = std::ptrdiff_t;
+    using pointer = value_type *;
+    using reference = value_type &;
+
+    user_iterator() = default;
+
+    bool operator==(const user_iterator &x) const { return UI == x.UI; }
+    bool operator!=(const user_iterator &x) const { return !operator==(x); }
+
+    // /// Return true if this iterator is at the end of the uses list.
+    bool atEnd() const { return *this == user_iterator(); }
```

Comment:
Do we need this? Seems like a weird thing for an iterator to provide. If we do, the implementation could just delegate to `UI.atEnd()`.

Reply:
Is it used somewhere?

Reply:
It may not be used anywhere. It exists in use_iterator and the IR equivalent of use/user_iterator. The implementation is based on IR's user_iterator which defers to use_iterator.

Reply:
I just noticed that IR only has this on user_iterator not on use_iterator. Not sure what its used for. Looks unused for SelectionDAG so I'll just delete.